Improve from db object perf

Deraen opened the pull request:
When querying data from larger data set (60 000 largeish documents) I noticed that converting DBObjects to maps takes large part of the run time.

Creating temporary sequence from DBObject is quite slow. Using .keySet for reduce collection and calling .get inside reduce is considerably faster. The common code between Map and DBObject implementations can't be shared as reflection would completely kill the performance and function can't be type hinted as DBObject doesn't implement Map interface.

I added a simple test case for the from-db-object performance. I'm seeing performance increase of 20% on this (170ms -> 140ms).

I also tried using transients but that doesn't seem to make any difference. If you have ideas how to share the implementation between Map and DBObject please tell. Looks like using `^Map` hint works, but I'm not sure if that's a good ideas as DBOject doesn't implement Map?

michaelklishin commented:
IIRC `DBObject` used to implement `java.util.Map`. If the hint is no longer applicable, we can either change it or remove it altogether. Feel free to do this in this PR, in fact.

Deraen commented:
Looks like I didn't look too well on Mongo API docs. `BasicDBObject` does implement Map: http://api.mongodb.org/java/current/com/mongodb/BasicDBObject.html

`DBObject` doesn't: http://api.mongodb.org/java/current/com/mongodb/DBObject.html

I wonder if it's a problem that it's possible that there could be some DBObjects that don't implement Map?

Deraen commented:
Also, looks like I pushed broken commit. Should be fixed now.

michaelklishin commented:
@Deraen either way, an incorrect type hint is more likely to prevent the JVM from optimising things than not having any tag in the first place. So we need to either remove or correct it.

Deraen commented:
Currently there are no type hints as the code is duplicated in both Map and DBObject implementations.

Is there use for the Map implementation anymore?

michaelklishin commented:
@Deraen the only way to check is to remove it and run the tests ;)

Deraen commented:
Seems to work, though I'm seeing something strange going on with tests in general. I already fixed two tests I at least thought where incorrect. Now I'm seeing other tests failing the same way: https://github.com/michaelklishin/monger/blob/master/test/monger/test/updating_test.clj#L47

`(is (= (doc (mc/find-by-id db collection doc-id))))`

What is going on here? It will always be true as `=` only has one parameter.

For some reason these cases with only one parameter to `=` throw Exception when running `lein test` but now when running `lein with-profile dev test`.

michaelklishin commented:
@Deraen that's an s-expression nesting overlook, there's no big idea.

Deraen commented:
Well, I did and just fixed those tests too. It exposed several broken tests  where DBObjects where compared to Clojure maps and I fixed those.

There's still a problem in `updating-test` where expected values have dates as `java.util.Date` and `from-db-object` returns `org.joda.time.DateTime`s. Perhaps easiest solution would be to remove those dates from test data in `updating-test`

michaelklishin commented:
@Deraen fair enough.

Deraen commented:
Found another fix: using `find-by-id` instead of `find-map-by-id` and manually creating `DBObject` from the expected value. It was a smaller change. Should be good now.

michaelklishin commented:
Thank you!

michaelklishin commented:
@Deraen please submit a PR that updates change log and credit yourself :)

Deraen commented:
Btw. the reason why I was getting exceptions from bad `=` forms is that [humane-test-output](https://github.com/pjstadig/humane-test-output) adds a check to `=` test expression handler, as it's quite easy mistake to make and can lead to tests doing nothing. It seems that `lein with-profile` disables it.
